### PR TITLE
Make arith rewriter take into account types

### PR DESCRIPTION
**Pull request by ajreynol**
This makes it so that the arithmetic rewriter and its utilities take into account Real/Int types.  This ensures that the rewriter preserves types when subtyping is eliminated.

### CODE REVIEW

File: src/theory/arith/rewriter/addition.cpp [hunk 1]
```diff
@@ -185,16 +186,16 @@ void addToSum(Sum& sum, TNode n, bool negate)
...
-Node collectSum(const Sum& sum)
+Node collectSum(const Sum& sum, const TypeNode& tn)
 {
-  if (sum.empty()) return mkConst(Rational(0));
+  if (sum.empty()) return mkConst(Rational(0), tn);
   Trace("arith-rewriter") << "Collecting sum " << sum << std::endl;
   // construct the sum as nodes.
   NodeBuilder nb(Kind::ADD);
-  for (const auto& s : sum)
+  for (const std::pair<const Node, RealAlgebraicNumber>& s : sum)
```
Comment (nafur) [inline src/theory/arith/rewriter/addition.cpp:198]: I'd rather keep `auto`... :)

Comment (nafur) [inline src/theory/arith/rewriter/addition.cpp:198]: I'd rather keep `auto`... :)

Comment (ajreynol) [reply src/theory/arith/rewriter/addition.cpp:198]: This use of auto makes the code hard to parse:  note `mkMultTerm` is overloaded. One cannot see what variant is being called unless the type is made explicit.

I would suggest either this change or we should not use name overloading on `mkMultTerm`, e.g. `mkMultTermRan`, `mkMultRational`.  However, I prefer having the type in the `for`.

**Comment by ajreynol** [comment]
Marking "do not merge" as the proposed solution with `ARITH_EQ` may not require these changes.
